mqtt/mqtt-broker/src/edgehub_auth.rs

```diff
@@ -0,0 +1,370 @@
+            password("qwerty123"),
+        )
+        .await;
+
+        assert!(result.is_err());
+    }
+
+    #[tokio::test]
+    async fn no_result_code_results_error() {
+        let _m = mock("POST", "/authenticate/")
+            .with_status(200)
+            .with_body(r#"{"identity":"somehub/somedevice","version": "2222-22-22"}"#)
+            .create();
+
+        let result = auth_request(
+            "somehub/somedevice/api-version=2018-06-30",
+            password("qwerty123"),
+        )
+        .await;
+
+        assert!(result.is_err());
+    }
+
+    #[tokio::test]
+    async fn password_cred_sends_password() {
+        let _m = mock("POST", "/authenticate/")
+            .with_status(200)
+            .match_body(Matcher::Regex("\"password\":\"qwerty123\"".to_string()))
+            .with_body(
+                r#"{"result": 200, "identity":"somehub/somedevice","version": "2020-04-20"}"#,
+            )
+            .create();
+
+        let result = auth_request(
+            "somehub/somedevice/api-version=2018-06-30",
+            password("qwerty123"),
+        )
+        .await;
+
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn cert_cred_sends_cert() {
+        let _m = mock("POST", "/authenticate/")
+            .with_status(200)
+            .match_body(Matcher::Regex("\"certificate\":\"MIIBL".to_string()))
+            .with_body(
+                r#"{"result": 200, "identity":"somehub/somedevice","version": "2020-04-20"}"#,
+            )
+            .create();
+
+        let credentials =
+            Credentials::ClientCertificate(Certificate::from(decode(CERT.to_string()).unwrap()));
+
+        let result = auth_request("somehub/somedevice/api-version=2018-06-30", credentials).await;
+
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn sends_user_and_version() {
+        let _m = mock("POST", "/authenticate/")
+            .with_status(200)
+            .match_body(Matcher::AllOf(vec![
+                Matcher::Regex("\"username\":\"somehub".to_string()),
+                Matcher::Regex("\"version\":\"2020-04-20".to_string()),
+            ]))
+            .with_body(
+                r#"{"result": 200, "identity":"somehub/somedevice","version": "2020-04-20"}"#,
+            )
+            .create();
+
+        let credentials =
+            Credentials::ClientCertificate(Certificate::from(decode(CERT.to_string()).unwrap()));
+
+        let result = auth_request("somehub/somedevice/api-version=2018-06-30", credentials).await;
+
+        assert!(result.is_ok());
+    }
+
+    async fn auth_request(
+        username: &str,
+        credentials: Credentials,
+    ) -> Result<Option<AuthId>, Error> {
+        let authenticator = EdgeHubAuthenticator {
+            url: mockito::server_url() + "/authenticate/",
+        };
+
+        authenticator
+            .authenticate(Some(username.to_string()), credentials)
```

Comment:
Usually I put method under test in the test method body and not hide it in some other function, and extract setup step in the separate functions instead.